Feature - Add movies to watchlist

sjacobs1 opened the pull request:
##Description
added functionality to add a movie to watchlist. Only basic functionality is implemented with a single property for now. Once functionality is approved I will add the rest of the properties and add UI to display saved movies.
Print statements are currently used to check if added movies are fetched and printed

##Screenshots/GIFs
<img width="366" alt="Screenshot 2024-05-16 at 2 33 18 PM" src="https://github.com/sjacobs1/Pet-Movie-Project-Repo/assets/162406787/d8600cb5-0069-4272-9987-61f92fec57d6">
<img width="369" alt="Screenshot 2024-05-16 at 2 33 50 PM" src="https://github.com/sjacobs1/Pet-Movie-Project-Repo/assets/162406787/d5b18239-e890-4a6e-8b19-f4cdddef90ac">


##Checklist
- [x] The code follows the project's coding standards.
- [ ] 100% Unit test coverage for View Models
- [x] All warnings Resolved.

sjacobs1 committed:
Feature - Add movies to watchlist

simekanidev commented on MovieAppPetProject/MovieAppPetProject/Watchlist/WatchlistViewModel.swift:
```diff
@@ -0,0 +1,22 @@
+    func fetchAndDisplayWatchlistItems() {
+        let watchlistItems = coreDataManager.getAllWatchlistItems()
+        for item in watchlistItems {
+            print("Original Title: \(item.originalTitle ?? "N/A")")
```
Do you still need this print?

sjacobs1 replied:
that print is there just to make sure that the movies added are actually being fetched, it does print so I can remove it when I start displaying the movies with UI

simekanidev replied:
Please add a warning to remove the print at a later stage then

sjacobs1 replied:
will do, or I'll just remove it since it is working

simekanidev replied:
Yeah I guess you can just remove it. For future you need to get into the habit of using your debugger instead of print statements. Let me know if you would like me to show you

sjacobs1 replied:
I would appreciate it please bro

simekanidev commented on MovieAppPetProject/MovieAppPetProject/CoreDataFiles/CoreDataManager.swift:
```diff
@@ -0,0 +1,46 @@
+    func createItem(movieDetails: MovieDetails) {
+        do {
+            guard let context = context else {
+                throw CoreDataError.noContext
```
This can stay on one line

sjacobs1 replied:
will change

simekanidev commented on MovieAppPetProject/MovieAppPetProject/CoreDataFiles/CoreDataManager.swift:
```diff
@@ -0,0 +1,46 @@
+    func getAllWatchlistItems() -> [WatchList] {
+        do {
+            guard let context = context else {
+                throw CoreDataError.noContext
```
Same here, One line

sjacobs1 replied:
understood

simekanidev commented on MovieAppPetProject/MovieAppPetProject/GetMovieDetailsByID/GetMovieDetails-ViewModel.swift:
```diff
@@ -11,6 +11,7 @@ class MovieDetailsViewModel {
     // MARK: - Variables
     private let movieDetailsRepository: MovieDetailsRepositoryType
+    private let coreDataManager = CoreDataManager()
     private var movieID: Int = 0
```
```suggestion
    private var movieID = 0
```

simekanidev commented on MovieAppPetProject/MovieAppPetProject/Watchlist/WatchlistViewModel.swift:
```diff
@@ -0,0 +1,22 @@
+class WatchlistViewModel {
+
+    // MARK: - Variable
+    private let coreDataManager = CoreDataManager()
```
Lets rather inject CodeData in the constructor of the viewModel

sjacobs1 replied:
will do

sjacobs1 committed:
code changes

small code format changes and used dependency injection for CoreData

kieran-woodrow commented on MovieAppPetProject/MovieAppPetProject/CoreDataFiles/CoreDataManager.swift:
```diff
@@ -0,0 +1,41 @@
+            guard let context = context else { throw CoreDataError.noContext }
+            return try context.fetch(WatchList.fetchRequest())
+        } catch {
+            print("Error fetching watchlist items: \(error)")
```
Please replace these with proper error handling in the next PR

sjacobs1 replied:
understood. will do

kieran-woodrow commented on MovieAppPetProject/MovieAppPetProject/CoreDataFiles/WatchList+CoreDataClass.swift:
```diff
@@ -0,0 +1,15 @@
+//  WatchList+CoreDataClass.swift
+//  MovieAppPetProject
+//
+//  Created by Sebastian Jacobs on 2024/05/14.
```
If these files are not needed manually, you can set it to not store in the project by ticking class definition instead of manual when creating them

sjacobs1 replied:
ok cool. can I do tick it and remove them now knowing they'll be safe? since they're already there

kieran-woodrow replied:
not sure what you're trying to say. Can you explain it again?

sjacobs1 replied:
sorry, what I mean is since the manual files are already there now, is safe / ok for me to remove the manual files and then just tick class definition for the entity so that it's in the background like you said and it will be ok? or is there a specific way of removing them without breaking anything.

kieran-woodrow replied:
yeah should be able to

kieran-woodrow commented on MovieAppPetProject/MovieAppPetProject/GetMovieDetailsByID/GetMovieDetails-ViewController.swift:
```diff
@@ -19,6 +19,11 @@ class MovieDetailsViewController: UIViewController {
     @IBOutlet private weak var status: UILabel!
 
+    // MARK: - IBAction
+    @IBAction func addToWatchlistTapped(_ sender: UIButton) {
```
private

sjacobs1 replied:
will change

kieran-woodrow commented on MovieAppPetProject/MovieAppPetProject/GetMovieDetailsByID/GetMovieDetails-ViewModel.swift:
```diff
@@ -11,7 +11,8 @@ class MovieDetailsViewModel {
     // MARK: - Variables
     private let movieDetailsRepository: MovieDetailsRepositoryType
-    private var movieID: Int = 0
+    private let coreDataManager = CoreDataManager()
```
viewModel should never know about core data manager.
Viewmodel should only know about repo. The repo knows about core data manager
Please adjust

sjacobs1 replied:
will adjust accordingly

kieran-woodrow commented on MovieAppPetProject/MovieAppPetProject/Watchlist/WatchlistViewModel.swift:
```diff
@@ -0,0 +1,23 @@
+class WatchlistViewModel {
+
+    // MARK: - Variable
+    private let coreDataManager: CoreDataManager
```
same as earlier comment. ViewModel should not know about core data manager

sjacobs1 replied:
understood

kieran-woodrow commented on MovieAppPetProject/MovieAppPetProject/CoreDataFiles/CoreDataManager.swift:
```diff
@@ -0,0 +1,41 @@
+//  Created by Sebastian Jacobs on 2024/05/14.
+//
+
+import Foundation
```
can remove foundation

sjacobs1 replied:
will do

kieran-woodrow commented on MovieAppPetProject/MovieAppPetProject/CoreDataFiles/CoreDataManager.swift:
```diff
@@ -0,0 +1,41 @@
+    let context = (UIApplication.shared.delegate as? AppDelegate)?.persistentContainer.viewContext
+
+    // MARK: - Functions
+    func getAllWatchlistItems() -> [WatchList] {
```
fetchAllWatchlistItems()

sjacobs1 replied:
will change

sjacobs1 committed:
refactor from pr comments

sjacobs1 committed:
Merge branch 'develop' into feature/watchlist

sjacobs1 committed:
watchlist funstionality

added fetch items functions and delete function. watchlist displayed in a table view

sjacobs1 committed:
moved string to constants

simekanidev commented on MovieAppPetProject/MovieAppPetProject/GetMovieDetailsByID/GetMovieDetails-Repository.swift:
```diff
@@ -9,12 +9,19 @@ import Foundation
 
 protocol MovieDetailsRepositoryType {
     func fetchMovieDetails(movieID: Int, completion: @escaping (Result<MovieDetails, CustomError>) -> Void)
```
Add Type alias for closure

sjacobs1 replied:
will do in cleanup

kieran-woodrow commented on MovieAppPetProject/MovieAppPetProject/Watchlist/WatchlistTableViewCell.swift:
```diff
@@ -0,0 +1,27 @@
+    private var watchlistItem: WatchList?
+
+    // MARK: - IBOutlets
+    @IBOutlet private weak var savedMovieTitle: UILabel!
```
new line

sjacobs1 replied:
will do in cleanup
